src/Voron/Data/Graphs/Hnsw.Parallel.cs

```diff
@@ -0,0 +1,835 @@
+     * 
+     * To start with, we aren't actually using parallel here to say threads. Instead, we re-wrote
+     * the algorithm using yield an enumerators. And we run it using a dedicated runner that consumes and
+     * execute all the interleaved (vs. concurrent) operations.
+     * 
+     * Whenever we need to do an expensive operation (such as loading vectors, or computing distances),
+     * we yield to the caller, giving a chance for the rest of the system to make forward progress while
+     * the task is completed in the background. 
+     * 
+     * That async operation is _not_ scheduled on a different thread. Instead, it is queued until all
+     * current operations are completed, then we check what pending work we have and start a batch 
+     * loads of all the vectors we need. The next step is to run the distance computation using the 
+     * thread pool. When that is completed, we can continue with the next step.
+     * 
+     * The idea is that we run N interleaved tasks, where N between 1..MaxConcurrentBatches, and in 
+     * each one of them, we pick an item to be inserted to the graph. We then run the HNSW until we
+     * need to do an expensive operation (which we'll offload to the thread pool if it is computation, or
+     * do a batch preload to amortise the costs of going to disk). At that point, we yield to *another*
+     * interleaved operation. By the time we hit the MaxConcurrentBatches, we gathered enough vectors to load and
+     * distances to compute that we can really start pumping through all the items. 
+     * 
+     * The key here is to batching of I/O for loading the vectors. See the runner for handling that 
+     * part of the process. Both NodePlacement and NodePlacementRunner are working very closely 
+     * together to achieve this work.
+     * 
+     * # Distance computation using the thread pool
+     * 
+     * Distance computation is expensive, and we want to run it in parallel. Each work item that we 
+     * send to the thread pool already had its vectors loaded by the batch process, so we can assume 
+     * that they are ready in memory. The work item compares a vector to a set of vectors (typically all 
+     * the edges of a particular node) and returns the shortest distance or the filtered set of edges.
+     * 
+     * We use the thread pool because:
+     * * There is a known limit to the amount of work we have (up to MaxConcurrentBatches), and it
+     *   cannot grow without bound. We won't cause thread pool starvation.
+     * * The amount of work for each item is well scoped and _short_. Under 0.5ms for each work item, 
+     *   so we won't cause a bottleneck in the thread pool.
+     * * We tested using a dedicated thread pool, but those performed significantly worse than the 
+     *   default .NET one. 
+     */
+    public partial class Registration
+    {
+        private int _nextNodeIndex;
+
+        public int MaxConcurrentBatches = 512;
```

Comment:
Missed setting this from RavenDB configuration.

Reply:
That commit is not a part of this PR. Does the other branch contain any other changes that should be included in this PR?

Reply:
Thanks for noticing, fix.